jicofo-common/src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.kt

```diff
@@ -0,0 +1,769 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.impl.protocol.xmpp
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
+import org.jitsi.jicofo.TaskPools.Companion.ioPool
+import org.jitsi.jicofo.xmpp.XmppProvider
+import org.jitsi.jicofo.xmpp.muc.ChatRoomAvModeration
+import org.jitsi.jicofo.xmpp.muc.ChatRoomListener
+import org.jitsi.jicofo.xmpp.muc.ChatRoomMember
+import org.jitsi.jicofo.xmpp.muc.ChatRoomMemberImpl
+import org.jitsi.jicofo.xmpp.muc.MemberRole
+import org.jitsi.jicofo.xmpp.muc.MemberRole.Companion.fromSmack
+import org.jitsi.jicofo.xmpp.sendIqAndGetResponse
+import org.jitsi.jicofo.xmpp.tryToSendStanza
+import org.jitsi.utils.MediaType
+import org.jitsi.utils.OrderedJsonObject
+import org.jitsi.utils.event.EventEmitter
+import org.jitsi.utils.event.SyncEventEmitter
+import org.jitsi.utils.logging2.createLogger
+import org.jivesoftware.smack.PresenceListener
+import org.jivesoftware.smack.SmackException
+import org.jivesoftware.smack.XMPPConnection
+import org.jivesoftware.smack.XMPPException
+import org.jivesoftware.smack.packet.ExtensionElement
+import org.jivesoftware.smack.packet.IQ
+import org.jivesoftware.smack.packet.Presence
+import org.jivesoftware.smack.packet.PresenceBuilder
+import org.jivesoftware.smackx.muc.MUCAffiliation
+import org.jivesoftware.smackx.muc.MUCRole
+import org.jivesoftware.smackx.muc.MultiUserChat
+import org.jivesoftware.smackx.muc.MultiUserChatManager
+import org.jivesoftware.smackx.muc.Occupant
+import org.jivesoftware.smackx.muc.ParticipantStatusListener
+import org.jivesoftware.smackx.muc.UserStatusListener
+import org.jivesoftware.smackx.muc.packet.MUCAdmin
+import org.jivesoftware.smackx.muc.packet.MUCInitialPresence
+import org.jivesoftware.smackx.muc.packet.MUCItem
+import org.jivesoftware.smackx.muc.packet.MUCUser
+import org.jxmpp.jid.EntityBareJid
+import org.jxmpp.jid.EntityFullJid
+import org.jxmpp.jid.Jid
+import org.jxmpp.jid.impl.JidCreate
+import org.jxmpp.jid.parts.Resourcepart
+import java.util.concurrent.ConcurrentHashMap
+import javax.xml.namespace.QName
+
+@SuppressFBWarnings(
+    value = ["JLM_JSR166_UTILCONCURRENT_MONITORENTER"],
+    justification = "We intentionally synchronize on [members] (a ConcurrentHashMap)."
+)
+class ChatRoomImpl(
+    private val xmppProvider: XmppProvider,
+    private val roomJid: EntityBareJid,
+    /** Callback to call when the room is left. */
+    private val leaveCallback: (ChatRoomImpl) -> Unit
+) : ChatRoom, PresenceListener {
+    private val logger = createLogger().apply {
+        addContext("room", roomJid.toString())
+    }
+
+    private val memberListener: MemberListener = MemberListener()
+
+    private val userListener = LocalUserStatusListener()
+
+    /** Listener for presence that smack sends on our behalf. */
+    private var presenceInterceptor: org.jivesoftware.smack.util.Consumer<PresenceBuilder>? = null
```

Comment:
Does this still need to be fully-qualified, since `leaveCallback` doesn't need to be declared as the other type of `Consumer`?